File: algolens-api/algolens/domain/portfolio/history_segments.py
"""Where a strategy's history restarts, and what that does to a chart.

A strategy that moves from one book to another keeps producing one continuous
equity curve, but the money behind it changed: the composition before the move
and the composition after it are different portfolios. Drawing them as one line
invites exactly the reading that is wrong -- that the whole line describes one
thing whose return can be measured end to end.

DECISION (2026-09-06, John): option B. The line breaks at the move. The history
is explicitly segmented, the break is labelled with the two books, and no window
figure is allowed to span it.

What this module does NOT do is rewrite the numbers. The equity values either
side of a break are the engine's and stay exactly as published; the segments
only say where one stops describing the same portfolio as the next.
"""

from datetime import date, datetime

BOOK_CHANGE = "book_change"


def _as_date(value):
    """A date from whatever the row carries, or None if it carries nothing."""
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    if isinstance(value, str) and value:
        try:
            return datetime.fromisoformat(value.replace("Z", "+00:00")).date()
        except ValueError:
            return None
    return None


def book_change_breaks(assignments):
    """The dates a strategy's history restarts, earliest first.

    Only a genuine move breaks anything. An assignment row with no
    ``from_portfolio_id`` is the strategy being placed in its first book, and a
    row whose from and to are the same book is a re-record of where it already
    was: neither changes the composition, so neither breaks the line.
    """
    breaks = {}
    for assignment in assignments or ():
        moved_from = assignment.get("from_portfolio_id")
        moved_to = assignment.get("to_portfolio_id")
        if not moved_from or not moved_to or moved_from == moved_to:
            continue

        when = _as_date(assignment.get("created_at"))
        if when is None:
            continue

        # Two moves on one day collapse to one break -- the reader cares that
        # the line stops describing the same book, not how many hops it took.
        # The last one written names the book the history continues into.
        breaks[when] = {
            "date": when.isoformat(),
            "fromPortfolioId": breaks.get(when, {}).get("fromPortfolioId", moved_from),
            "toPortfolioId": moved_to,
            "reason": BOOK_CHANGE,
        }

    return [breaks[when] for when in sorted(breaks)]


def split_at_breaks(points, breaks):
    """``points`` grouped into segments, one per uninterrupted stretch of book.

    A point dated on or after a break belongs to the segment the break opens:
    the move takes effect that day, so that day's equity is the first of the new
    book's history rather than the last of the old one's.

    Always returns at least one segment, so a caller can render the result
    without special-casing "no breaks" or "no points".
    """
    boundaries = [
        _as_date(entry.get("date")) for entry in (breaks or ()) if entry.get("date")
    ]
    boundaries = sorted(b for b in boundaries if b is not None)

    segments = [[] for _ in range(len(boundaries) + 1)]
    for point in points or ():
        when = _as_date(point.get("date"))
        index = 0
        if when is not None:
            while index < len(boundaries) and when >= boundaries[index]:
                index += 1
        segments[index].append(point)

    return segments


def latest_segment(points, breaks):
    """The most recent unbroken stretch: what a window figure may measure.

    A return quoted over a window that spans a book change is a return on two
    different portfolios added together, which is not a number about anything.
    Empty only when there are no points at all.
    """
    segments = split_at_breaks(points, breaks)
    for segment in reversed(segments):
        if segment:
            return segment
    return []

File: algolens-frontend/src/domain/portfolio/historySegments.ts
/**
 * Where a strategy's equity curve stops describing the same portfolio.
 *
 * A strategy that moves between books keeps producing one continuous curve,
 * but the money behind it changed. Drawn as a single line it invites the one
 * reading that is wrong: that the whole line is one thing, measurable end to
 * end.
 *
 * Decision (2026-09-06): the line breaks at the move, the break is labelled
 * with the two books, and no window figure spans it. This mirrors
 * `algolens/domain/portfolio/history_segments.py`, which decides WHERE the
 * breaks are; this file only decides how they are drawn.
 */

export interface HistoryBreak {
  /** YYYY-MM-DD. The first day of the new book's history. */
  date: string;
  fromPortfolioId: string;
  toPortfolioId: string;
  reason: string;
}

interface Dated {
  date: string;
}

/** The calendar day of a stamp, as a sortable number. NaN if unreadable. */
function dayStart(date: string): number {
  return Date.parse(String(date).slice(0, 10));
}

/**
 * `points` grouped into segments, one per uninterrupted stretch of book. A
 * point dated on or after a break opens the next segment: the move takes
 * effect that day, so that day's equity is the new book's first, not the old
 * book's last.
 *
 * Always returns at least one segment.
 */
export function splitAtBreaks<T extends Dated>(
  points: readonly T[],
  breaks: readonly HistoryBreak[] | undefined
): T[][] {
  const boundaries = (breaks ?? [])
    .map(b => dayStart(b.date))
    .filter(d => !Number.isNaN(d))
    .sort((a, b) => a - b);

  const segments: T[][] = Array.from({ length: boundaries.length + 1 }, () => []);
  for (const point of points ?? []) {
    const day = dayStart(point.date);
    let index = 0;
    if (!Number.isNaN(day)) {
      while (index < boundaries.length && day >= boundaries[index]) index += 1;
    }
    segments[index].push(point);
  }
  return segments;
}

/**
 * The most recent unbroken stretch -- what a window return may be measured
 * over. A return spanning a book change is a return on two different
 * portfolios added together, which is not a number about anything.
 */
export function latestSegment<T extends Dated>(
  points: readonly T[],
  breaks: readonly HistoryBreak[] | undefined
): T[] {
  const segments = splitAtBreaks(points, breaks);
  for (let i = segments.length - 1; i >= 0; i -= 1) {
    if (segments[i].length > 0) return segments[i];
  }
  return [];
}

/**
 * The points a chart should plot: the same series with a null-valued spacer
 * between segments, which is how recharts is told to lift the pen. Charts
 * must not set `connectNulls`, or the spacer is drawn straight over.
 */
export function withBreakGaps<T extends Dated & { value: number | null }>(
  points: readonly T[],
  breaks: readonly HistoryBreak[] | undefined
): (T | { date: string; value: null })[] {
  const segments = splitAtBreaks(points, breaks).filter(s => s.length > 0);
  if (segments.length < 2) return [...points];

  const plotted: (T | { date: string; value: null })[] = [];
  segments.forEach((segment, index) => {
    // The spacer carries the day before the segment opens, so it sits in the
    // gap rather than on top of the first real point of the new book.
    if (index > 0) {
      const opensAt = dayStart(segment[0].date);
      const spacerDay = Number.isNaN(opensAt)
        ? segment[0].date
        : new Date(opensAt - 86_400_000).toISOString().slice(0, 10);
      plotted.push({ date: spacerDay, value: null });
    }
    plotted.push(...segment);
  });
  return plotted;
}

/** The breaks that fall inside a window, so only visible ones are labelled. */
export function breaksWithin<T extends Dated>(
  points: readonly T[],
  breaks: readonly HistoryBreak[] | undefined
): HistoryBreak[] {
  if (!breaks?.length || !points?.length) return [];
  const days = points.map(p => dayStart(p.date)).filter(d => !Number.isNaN(d));
  if (!days.length) return [];
  const first = Math.min(...days);
  const last = Math.max(...days);
  return breaks.filter(b => {
    const day = dayStart(b.date);
    return !Number.isNaN(day) && day > first && day <= last;
  });
}
